### PR TITLE
Safes, Torches, and Gas Handling, oh my! [WIP]

### PR DESCRIPTION
**Pull request by i2amroy**
So this originally started out as a PR to finish off player interactions with safes, but upon noticing that it was currently impossible to make it take several hours to cut through a safe with an acetylene torch I ended up expanding it's breadth to include basic gas handling as well.

**Things in this PR:**
1. Adds the ability for players to input a random number into a combination safe (most important thing right there! :P)
2. Adds the ability for players to store gases inside of any size of metal tank (or future containers with the "airtight" flag). Currently the only gases are hydrogen and an acetylene gas mixture used in acetylene torches. If the gas is dumped out it immediately escapes. The vast majority of this uses the liquid handling code, which remains virtually unchanged (meaning that "liquid" in variable names often now means a more broad definition of "liquid", as in liquids and vapors).
3. Hydrogen and acetylene now both spawn contained in metal tanks.
4. Cut through safes with acetylene torches. This time varies, but takes on average about 4 hours (done as a one_in(10), since the torch can only hold about 24 minutes worth of gas at once).
5. Siphon hydrogen from vehicles.
6. Refill acetylene torches.
7. Cracking safes now is a long action, trains the mechanics skill, and depends on the mechanics skill to determine how long it takes. With higher skill it can be a quite short amount of time, 20-30 minutes, but with extremely low skills it might be faster just to cut your way into the safe.
8. Gave acetylene torches the ability to occasionally spawn with the smaller tools instead of just the bigger ones in hardware stores.
9. Acetylene refills now spawn most places the torches do.
10. Hydrogen now spawns in a few more lab-like places (such as school labs).
11. Since we currently have real hydrogen-powered cars coming onto the market, moved the plasma (hydrogen) engine from the vault spawn list to the mechanics shop. It's still fairly rare, but it's a much more common location now.
12. The player can now make hydrogen from clean water and a chemistry set, essentially electrolyzing water with no cost other than the water itself and the batteries to power the set.

**Drawbacks**
Any hydrogen left sitting on the ground (not in an inventory, weapon, or car) will become unable to be picked up. I didn't really consider this that big of a drawback because it used to be a fairly end-game item and is becoming a _much_ more common one (level 3 crafting recipe); so pretty much everyone with stocks of hydrogen can afford to toss away an hour in-game to totally recraft their entire lost stocks and then some. This was mainly due to the difficulty of casing items in something else due to the way we process items from save loading (storing the container and what it contains, rather than the item and what contains it).

**Things to look out for**
I might have missed a gas vs liquid handling somewhere. I'm pretty sure I tested all of the new places where you could use a gas instead of a liquid, but it's something to check out.

**Possible future PR's**
1. Now that hydrogen cars are becoming more popular in the real world, it might be worthwhile to come up with a car design that uses a plasma engine.
2. Hydrogen stations? Similar to a gas station, it would give us a place to put hydrogen in bulk without the player crafting it.
3. Maybe have a car lot where we can stick a bunch of different car designs and some places to stick some of the rarer engine types such as hydrogen.
4. Shift oxygen over to the new system, let players refill oxygen tanks and craft it through the same process as hydrogen. Make players need to craft oxygen and acetylene together to refill acetylene torches (rather than hand-waving it as a "mixture").

### CODE REVIEW

**Comment by kevingranade** [comment]
Hydrogen production via electrolysis is right out, the problem isn't making
the gas, but loading it into the fuel cells, which must have it in some
bizarre near-liquid yet somehow stable state, this requires either immense
pressure or some kind of techno magic.
If you want to support gaseous hydrogen use, you can produce with
electrolysis, then run it through a compressor feeding a high-pressure tank
that holds less energy/lb than a gas tank, AFAIK that's the extent of
gaseous hydrogen storage.
A fuel cell regenerator in a lab or industrial setting that you can feed
power to create hydrogen would be fine, but it would be a lot of power.

**Comment by i2amroy** [comment]
I think your knowledge might be a bit behind current technology here. Many hydrogen fuel cell designs today just literally bubble in pure hydrogen and oxygen through two different carbon electrodes, no fancy pseudo-liquids or anything like that. We're also getting better at developing low pressure alternatives; some designs today can take hydrogen and oxygen at pressures as low as 10 atm. I really don't think it's that much of a stretch to see player produced hydrogen being used in a vehicle. 

Though I *can* see requiring a pump as a tool to be used in crafting it to get it up to pressure at the start, as well as siphoning it out of a car. (Requiring a pump just to move it from one container to another would also be something that could be cool, but unlike those first two it would require ripping apart a *lot* of code, and probably would be better off in a follow-up PR).

**Comment by Coolthulhu** [comment]
The gas idea itself looks good, but:
* Instead of checking for "is gas or liquid", it would probably be better to just check for solid instead. Harder to make a mistake and less lines of code. Plus, if any one wanted to make use of the PLASMA phase (unlikely, but still), it would be handled already.
* Moving gas from a container to container probably shouldn't be allowed without tools, at least not from a large container to a smaller one. Alternatively, they should only be equalized - ie. when moving from a 100/400 container to a 0/10 one, only 3-4 units of gas volume are transfered.

We had a discussion about gaseous hydrogen on forums. Hydrogen is notoriously hard to store, especially under pressure.
I recall Kevin suggesting a separate "hydrogen-tight" tank, not craftable by the player.

**Comment by i2amroy** [comment]
>Instead of checking for "is gas or liquid", it would probably be better to just check for solid instead. Harder to make a mistake and less lines of code. Plus, if any one wanted to make use of the PLASMA phase (unlikely, but still), it would be handled already.

Fair enough.

>Moving gas from a container to container probably shouldn't be allowed without tools, at least not from a large container to a smaller one. Alternatively, they should only be equalized - ie. when moving from a 100/400 container to a 0/10 one, only 3-4 units of gas volume are transfered.

Honestly I want to do this, but to do so would involve ripping apart a very large part of the handling code (and thus why I didn't do it). Right now the code for gases pretty much just uses the straight liquid code, which cuts down drastically on the changes.

>Hydrogen is notoriously hard to store, especially under pressure.

As I mentioned just a moment ago, we're getting better and better at low pressure alternatives to fuel cells if need be. In addition a very large part of our current hydrogen tank design isn't necessarily caused by the need to hold hydrogen; it's because our current safety standards require the tank to resist everything up to high-caliber close range gun fire. The original basic steel tanks were still fine at holding hydrogen up to around 200 atm no problem.

**Comment by i2amroy** [comment]
Had an idea of how I might possible implement requiring the new pump item I made for simple container -> container transfers of gases. Don't have time right now, but will look into it either much later tonight or tomorrow as well as finishing off the other pump changes (which are definitely looking more promising). With any luck I should be able to have pump charges required for any operation involving the moving of gas from one container to another.

**Comment by kevingranade** [comment]
The problem isn't the pressure of the hydrogen while being used, the problem is the pressure while being stored.

Based on Sources: https://en.wikipedia.org/wiki/Energy_density
https://en.wikipedia.org/wiki/Compressed_hydrogen
Specifically the following facts:
Hydrogen at at least 5,000psi has an energy density of 142 Mj/kg or 5.6 Mj/L (This is the lower boundary for what is considered "Compressed Hydrogen", I'm taking it as the representative pressure, though most likely the figures actually refer to 10,000 PSI Hydrogen, in which case all the numbers are 2x worse for Hydrogen.)
Liquid diesel has an energy density of 48 Mj/kg or 35.8 Mj/L
We can draw some inferences.
Survivor-achievable pressures for gaseous hydrogen, say 500psi or so (this is really high, it requires a rather large pump to get pressures like this, and a "survivor special" is going to have serious problems achieving it), are going to have an energy density of 0.11 Mj/L, so for a "tank of gas" (assuming 64L, seems typical), you're going to need a container that holds 22,400 L of hydrogen.  The weight of the hydrogen is inconsequential at that pressure, but the tank is going to be enormously heavy, not to mention simply huge.
You can see how that applies to the weaponized Hydrogen items as well, it's simply not reasonable to carry around enough Hydrogen to do any meaningful work at that pressure.

**Comment by i2amroy** [comment]
I'll give you that the tanks won't hold at the pressure needed (which, of course, has a simple solution of having rarer special tanks needed to store the high-pressured gases). And of course it's not going to work as a "survivor special". This isn't going to be a pump you can cobble together from some spare rubber hose and a bit of string. *However*, the current designs for a lot of hydrogen compressors for home use are nowhere near as big as you think they are. Check this out: https://en.wikipedia.org/wiki/Hydrogen_station
As noted at the bottom, in 2007 Australia developed a home hydrogen system that electrolyzed and compressed hydrogen ready to be used; the whole system is about the size of a filing cabinet. [Here's](http://www.hydrogengas.biz/images/page27image.jpg) a picture of the Honda Home Energy Station, currently in development, which does the same thing.

This might not be something that you could make by yourself, but this is certainly something that you could pick up somewhere and mount on your car to be used (ala the Foodco buddy).

So here's what I propose:
1. Strike the hydrogen recipe. Having a compressed and uncompressed version of hydrogen would just be confusing.
2. Remove the pump item I'm currently working on (since a real pump for those pressures would be about as big as a filing cabinet). Gas can't be transferred from tank to tank, but *can* be unloaded into the atmosphere if you need to empty a tank for some reason or another.
3. Metal tanks are no longer valid for holding gases. Instead we have a special "compressed gas tank" that can hold gases. Hydrogen tanks on vehicles now require the new tank.
4. New furniture type, the "Hydrogen Pump". Shows up at a new gas station type: Hydrogen Station. Maybe stick it in a possible house design or two as well. The Hydrogen Pump works identical to a gas pump, but holds hydrogen instead of gasoline. No water being pumped in though, so the station won't refill it's hydrogen amounts on it's own (which is a nice lore explanation for why it isn't filling up despite having solar panels).
5. Hydrogen Pumps can be disassembled. On disassembly each yields a good solar panel (not a quantum one), a few (2?) of the special tanks, and a "Hydrogen Pump" (Item).
6. The "Hydrogen Pump" item can be installed into a car. When turned on, it draws electricity from the car battery and clean water from an installed storage tank and produces hydrogen in the car's hydrogen tanks.
7. Hydrogen can then be siphoned out of the car into the aforementioned tanks, which can be used for anything.
8. Possible new car design: "The Mirage" (Poking fun at the [Mirai](https://en.wikipedia.org/wiki/Toyota_Mirai)). Runs off of a hydrogen-based engine.
9. Acetylene stays as it currently is in the PR, but uses the new gas tanks.

**Comment by tyrael93** [comment]
it absolutely makes sense that hydrogen station would be more common in 2040something since we're starting production right now too. Hell, wouldn't regular gasoline be rarer than hydrogen by then? I remember reading that at the very best we'll be pumping out the last drop of oil by 2050, while the only thing you need for hydrogen is electrolysis.

Would making a mod to replace every car's engine with a hydrogen-based one be difficult?

**Comment by kevingranade** [comment]
PR handling issue, it'd be good to split off the hydrogen stuff from the acetylene stuff, the latter is totally awesome and can just go in :)

I disagree that having two hydrogens is a problem, and it forces us into an uncomfortable middle ground that doesn't suit either the DIY end of the scale or the super-science end of the scale.  To make it clear though, how about we rename the current "hydrogen" to "Tritium" or "Deuterium", which fit the fusion theme better than hydrogen anyway?
That clears the way to make a modern fuel type hydrogen without mixing it up with super science.  To be clear the thing I'm objecting to is mixing those two, not to having a hydrogen as a fuel at all.  As far as that goes I just want it to be an accurate portrayal, which means separating out the plasma bits.

**Comment by i2amroy** [comment]
Fair enough, I'll look into reverting some of the hydrogen stuff from this and then setting up a new PR (that will, of course, depend on this being merged in, since it'll be pulling the gas handling stuff out).

File: src/advanced_inv.cpp [hunk 1]
```diff
@@ -1789,7 +1789,13 @@ bool advanced_inventory::add_item( aim_location destarea, item &new_item )
...
 {
-    if( !src_container.is_watertight_container() ) {
+    item &src = src_container.contents[0];
+
+    if( src.made_of( LIQUID ) && !src_container.is_watertight_container()) {
+        popup( _( "Source must be watertight container." ) );
+        return false;
+    }
+    if( src.made_of( GAS ) && !src_container.is_airtight_container()) {
         popup( _( "Source must be watertight container." ) );
```
Comment (BevapDin) [inline src/advanced_inv.cpp]: > "Source must be watertight container."

Than why are you checking for an airtight container?

File: src/iexamine.cpp [hunk 2]
```diff
@@ -974,23 +974,35 @@ void iexamine::slot_machine(player *p, map*, const tripoint&)
 void iexamine::safe(player *p, map *m, const tripoint &examp)
 {
     if (!p->has_amount("stethoscope", 1)) {
-        add_msg(m_info, _("You need a stethoscope for safecracking."));
-        return;
+        if (query_yn(_("The safe is locked. Input a random combination?"))) {
```
Comment (BevapDin) [inline src/iexamine.cpp]: This action does not seem to consume any moves, I could just make it again and again (using a auto-type program) until it succeeds. It should at least take some seconds.

**Comment by BevapDin** [comment]
> Instead of checking for "is gas or liquid", it would probably be better to just check for solid instead. Harder to make a mistake and less lines of code.

You want to have specific function that check named requirements, not for specific properties of the item. Let the item itself decide if those requirements are true.

- `item::needs_to_be_in_container` (true for liquids and gases, triggers calling `handle_liquid` instead of just putting it into the inventory)
- `item::can_grabbed` (can be picked up from ground, false for liquids and gases outside of containers)

Using those functions makes it clear what the intend is and requires far less changes here.

----

You should PR the safe cracking and the gas handling in separate PRs, they have nothing in common.

**Comment by i2amroy** [comment]
>You should PR the safe cracking and the gas handling in separate PRs, they have nothing in common.

Well, they're both required for new ways to open safes in different ways :P. I guess I could put the safe cracking in one (which would be like a 20 line PR), the reloadable gas system in another (which is required for all for all the acetylene torch changes, because else it's impossible to get realistic cutting times on opening safes), and the hydrogen content in a third one that depends on the second landing.

**Comment by Frost-wood** [comment]
Maybe as a compromise we could have low-grade hydrogen used for fuel cells and other uses and a high-grade hydrogen used for plasma engines, and weapons.

You could even give the higher-grade hydrogen a corporate name.

Alternatively it the plasma-engine could be based on the VASIMR engine proposed for spacecraft, using argon or xenon instead of hydrogen.  Thus lessening the confusion between the two types of fuel.

**Comment by i2amroy** [comment]
Closing this in light of reworks and PR splitting.

**Pull request closed without merging**